File: src/agents/pacman.py
import random
import pygame
import math
from typing import Tuple, List
from ..config.constants import Direction, CELL_SIZE, PACMAN_SPEED, CellType
from .base_agent import BaseAgent
from ..algorithms.search import AStarSearch, BreadthFirstSearch, UniformCostSearch

class PacmanAgent(BaseAgent):
    def __init__(self, x: int, y: int):
        super().__init__(x, y)
        self.score = 0
        self.is_powered_up = False
        self.power_timer = 0
        self.mouth_angle = 0
        self.mouth_speed = 5
        self.opening_mouth = True
        
        # AI components
        self.current_path = []
        self.search_algorithm = AStarSearch()
        self.autonomous_mode = True
        self.stuck_timer = 0
        self.last_position = (x, y)
        self.no_movement_counter = 0
        
        # Colors
        self.YELLOW = (255, 255, 0)
        self.BLACK = (0, 0, 0)
    
    def get_pellet_positions(self, maze) -> List[Tuple[int, int]]:
        """Get positions of all pellets in the maze"""
        pellets = []
        for y in range(maze.height):
            for x in range(maze.width):
                cell_type = maze.get_cell_type(x, y)
                if cell_type in [CellType.PELLET, CellType.POWER_PELLET]:
                    pellets.append((x, y))
        return pellets
    
    def get_next_move(self, maze) -> Direction:
        """Get the next move based on AI or manual control"""
        if not self.autonomous_mode:
            return self._get_manual_input()
        
        return self._get_ai_move(maze)
    
    def _get_manual_input(self) -> Direction:
        """Handle manual keyboard input"""
        keys = pygame.key.get_pressed()
        
        if keys[pygame.K_UP]:
            return Direction.UP
        elif keys[pygame.K_DOWN]:
            return Direction.DOWN
        elif keys[pygame.K_LEFT]:
            return Direction.LEFT
        elif keys[pygame.K_RIGHT]:
            return Direction.RIGHT
        elif keys[pygame.K_SPACE]:
            self.toggle_control_mode()
        
        return self.direction
    
    def _get_ai_move(self, maze, ghost_positions: List[Tuple[int, int]]) -> Direction:
        """Calculate next move using AI with ghost avoidance"""
        current_pos = (int(round(self.x)), int(round(self.y)))
        
        # Check if we're in danger (too close to ghosts)
        in_danger = False
        for ghost_pos in ghost_positions:
            distance = abs(current_pos[0] - ghost_pos[0]) + abs(current_pos[1] - ghost_pos[1])
            if distance < 3:  # Danger threshold
                in_danger = True
                break
        
        # If in danger, focus on escaping
        if in_danger and not self.is_powered_up:
            # Find direction that maximizes distance from all ghosts
            best_direction = self.direction
            max_min_distance = -1
            
            for direction in [Direction.RIGHT, Direction.LEFT, Direction.UP, Direction.DOWN]:
                next_x = current_pos[0] + direction.value[0]
                next_y = current_pos[1] + direction.value[1]
                
                if maze.is_valid_position(next_x, next_y):
                    # Calculate minimum distance to any ghost from this position
                    min_ghost_distance = min(
                        abs(next_x - gx) + abs(next_y - gy)
                        for gx, gy in ghost_positions
                    )
                    
                    if min_ghost_distance > max_min_distance:
                        max_min_distance = min_ghost_distance
                        best_direction = direction
            
            self.current_path = []  # Clear current path when in danger
            return best_direction
        
        # Normal pathfinding to pellets when safe
        if not self.current_path or self._is_near_target():
            pellets = self.get_pellet_positions(maze)
            if pellets:
                # Find the safest pellet to target
                safest_pellet = max(pellets, key=lambda p: min(
                    abs(p[0] - gx) + abs(p[1] - gy)
                    for gx, gy in ghost_positions
                ))
                self.current_path = self.search_algorithm.find_path(
                    current_pos, [safest_pellet], maze)
        
        # If we have a path, follow it
        if self.current_path:
            target = self.current_path[0]
            dx = target[0] - self.x
            dy = target[1] - self.y
            
            if abs(dx) > abs(dy):
                return Direction.RIGHT if dx > 0 else Direction.LEFT
            else:
                return Direction.DOWN if dy > 0 else Direction.UP
        
        # If no path, find a safe direction
        valid_directions = []
        for direction in [Direction.RIGHT, Direction.LEFT, Direction.UP, Direction.DOWN]:
            next_x = current_pos[0] + direction.value[0]
            next_y = current_pos[1] + direction.value[1]
            if maze.is_valid_position(next_x, next_y):
                # Check if this direction is safe from ghosts
                min_ghost_distance = min(
                    abs(next_x - gx) + abs(next_y - gy)
                    for gx, gy in ghost_positions
                )
                if min_ghost_distance >= 2:  # Safe distance threshold
                    valid_directions.append(direction)
        
        if valid_directions:
            return random.choice(valid_directions)
        
        # If no safe direction, keep current direction if valid
        next_x = current_pos[0] + self.direction.value[0]
        next_y = current_pos[1] + self.direction.value[1]
        if maze.is_valid_position(next_x, next_y):
            return self.direction
            
        # Last resort: any valid direction
        for direction in [Direction.RIGHT, Direction.LEFT, Direction.UP, Direction.DOWN]:
            next_x = current_pos[0] + direction.value[0]
            next_y = current_pos[1] + direction.value[1]
            if maze.is_valid_position(next_x, next_y):
                return direction
                
        return self.direction
    
    def _is_near_target(self) -> bool:
        """Check if Pacman is near the next target in the path"""
        if not self.current_path:
            return True
            
        target = self.current_path[0]
        distance = math.sqrt((self.x - target[0])**2 + (self.y - target[1])**2)
        return distance < 0.3  # Increased threshold
    
    def update(self, maze, ghost_positions: List[Tuple[int, int]]):
        """Update Pacman's position and state"""
        # Get next move from AI or manual control
        if self.autonomous_mode:
            next_direction = self._get_ai_move(maze, ghost_positions)
        else:
            next_direction = self._get_manual_input()
        
        # Calculate next position
        next_x = self.x + next_direction.value[0] * PACMAN_SPEED
        next_y = self.y + next_direction.value[1] * PACMAN_SPEED
        
        # Round positions for better grid alignment
        cell_x = int(round(next_x))
        cell_y = int(round(next_y))
        
        # Check if the move is valid
        if maze.is_valid_position(cell_x, cell_y):
            # Check if move is safe from ghosts
            if self.autonomous_mode and not self.is_powered_up:
                too_close_to_ghost = any(
                    abs(cell_x - gx) + abs(cell_y - gy) < 2
                    for gx, gy in ghost_positions
                )
                if too_close_to_ghost:
                    return  # Don't make the move if it's too dangerous
            
            self.x = next_x
            self.y = next_y
            self.direction = next_direction
            
            # Remove reached positions from path
            if self.current_path and self._is_near_target():
                self.current_path.pop(0)
            
            # Check for pellet collection
            cell_x = int(round(self.x))
            cell_y = int(round(self.y))
            
            if (abs(self.x - cell_x) < 0.3 and abs(self.y - cell_y) < 0.3):
                is_power_pellet = maze.eat_pellet(cell_x, cell_y)
                if is_power_pellet:
                    self.is_powered_up = True
                    self.power_timer = 600
                    self.score += 50
                else:
                    self.score += 10
        
        # Update power-up timer
        if self.is_powered_up:
            self.power_timer -= 1
            if self.power_timer <= 0:
                self.is_powered_up = False
        
        # Update mouth animation
        if self.opening_mouth:
            self.mouth_angle = min(45, self.mouth_angle + self.mouth_speed)
            if self.mouth_angle >= 45:
                self.opening_mouth = False
        else:
            self.mouth_angle = max(0, self.mouth_angle - self.mouth_speed)
            if self.mouth_angle <= 0:
                self.opening_mouth = True
    
    def draw(self, screen, offset_y=0):
        """Draw Pacman with vertical offset"""
        # Calculate center position
        center_x = int(self.x * CELL_SIZE + CELL_SIZE // 2)
        center_y = int(self.y * CELL_SIZE + CELL_SIZE // 2) + offset_y
        
        direction_angle = {
            Direction.RIGHT: 0,
            Direction.UP: 90,
            Direction.LEFT: 180,
            Direction.DOWN: 270
        }[self.direction]
        
        # Draw Pacman body
        pygame.draw.circle(screen, self.YELLOW, (center_x, center_y), 
                         int(CELL_SIZE * 0.8 // 2))
        
        # Draw mouth
        if self.mouth_angle > 0:
            start_angle = math.radians(direction_angle - self.mouth_angle)
            end_angle = math.radians(direction_angle + self.mouth_angle)
            
            pygame.draw.arc(screen, self.BLACK, 
                          (center_x - CELL_SIZE//2, 
                           center_y - CELL_SIZE//2,
                           CELL_SIZE, CELL_SIZE),
                           start_angle, end_angle, 3)
        
        # Draw the planned path
        if self.autonomous_mode and self.current_path:
            for x, y in self.current_path:
                path_center_x = int(x * CELL_SIZE + CELL_SIZE // 2)
                path_center_y = int(y * CELL_SIZE + CELL_SIZE // 2) + offset_y
                pygame.draw.circle(screen, (255, 0, 0), 
                                (path_center_x, path_center_y), 2)
    
    def toggle_control_mode(self):
        """Toggle between AI and manual control"""
        self.autonomous_mode = not self.autonomous_mode
        self.current_path = []
        print(f"Control mode: {'AI' if self.autonomous_mode else 'Manual'}")

File: web/js/agents/pacman.js
import { BaseAgent } from './baseAgent.js';
import { PACMAN_SPEED, CELL_SIZE, Direction, CellType } from '../config/constants.js';
import { AStarSearch } from '../core/pathfinding.js';

export class Pacman extends BaseAgent {
    constructor(x, y) {
        super(x, y);
        this.score = 0;
        this.isPoweredUp = false;
        this.powerTimer = 0;
        this.mouthAngle = 0;
        this.mouthSpeed = 5;
        this.openingMouth = true;
        this.speed = PACMAN_SPEED;
        
        // AI components
        this.currentPath = [];
        this.searchAlgorithm = new AStarSearch();
        this.autonomousMode = true;
        this.dangerThreshold = 3;
        this.pathUpdateCooldown = 0;
        this.stuckDetectionThreshold = 10;
        this.lastPosition = { x, y };
        this.noMovementCounter = 0;
        this.color = '#FFFF00';
        this.direction = Direction.RIGHT; // Default direction
    }

    reset() {
        this.resetSpeed();
        this.currentPath = [];
        this.pathUpdateCooldown = 0;
        this.noMovementCounter = 0;
        this.lastPosition = { x: this.x, y: this.y };
        this.isPoweredUp = false;
        this.powerTimer = 0;
        this.direction = Direction.RIGHT;
        this.score = 0;
    }

    handleInput(key) {
        if (!this.autonomousMode) {
            const directionMap = {
                'ArrowUp': Direction.UP,
                'ArrowDown': Direction.DOWN,
                'ArrowLeft': Direction.LEFT,
                'ArrowRight': Direction.RIGHT
            };
            
            if (directionMap[key]) {
                const nextX = Math.round(this.x) + directionMap[key].x;
                const nextY = Math.round(this.y) + directionMap[key].y;
                if (this.maze && this.maze.isValidPosition(nextX, nextY)) {
                    this.direction = directionMap[key];
                }
            }
        }
    }

    isStuck() {
        const currentPos = { x: Math.round(this.x), y: Math.round(this.y) };
        
        if (Math.abs(currentPos.x - this.lastPosition.x) < 0.1 && 
            Math.abs(currentPos.y - this.lastPosition.y) < 0.1) {
            this.noMovementCounter++;
        } else {
            this.noMovementCounter = 0;
        }

        this.lastPosition = { ...currentPos };
        return this.noMovementCounter > this.stuckDetectionThreshold;
    }

    calculateGhostDanger(pos, ghostPositions) {
        if (!ghostPositions.length) return Infinity;
        
        return Math.min(...ghostPositions.map(ghost => 
            Math.abs(pos.x - ghost.x) + Math.abs(pos.y - ghost.y)
        ));
    }

    findBestPellet(pellets, ghostPositions, maze) {
        if (!pellets.length) return null;

        return pellets.reduce((best, pellet) => {
            const ghostDanger = this.calculateGhostDanger(pellet, ghostPositions);
            const distanceToMe = Math.abs(pellet.x - this.x) + Math.abs(pellet.y - this.y);
            const escapeRoutes = this.countEscapeRoutes(pellet.x, pellet.y, maze);
            const isPowerPellet = maze.getCellType(pellet.x, pellet.y) === CellType.POWER_PELLET;
            
            const score = (ghostDanger * 2) - (distanceToMe * 0.5) + 
                         (escapeRoutes * 0.3) + (isPowerPellet ? 10 : 0);
            
            if (!best || score > best.score) {
                return { pellet, score };
            }
            return best;
        }, null).pellet;
    }

    getPelletPositions(maze) {
        const pellets = [];
        for (let y = 0; y < maze.height; y++) {
            for (let x = 0; x < maze.width; x++) {
                const cellType = maze.getCellType(x, y);
                if (cellType === CellType.PELLET || cellType === CellType.POWER_PELLET) {
                    pellets.push({ x, y });
                }
            }
        }
        return pellets;
    }

    findEscapeDirection(currentPos, ghostPositions, maze) {
        let bestDirection = this.direction;
        let maxSafety = -Infinity;

        for (const direction of Object.values(Direction)) {
            const nextX = currentPos.x + direction.x;
            const nextY = currentPos.y + direction.y;

            if (!maze.isValidPosition(nextX, nextY)) continue;

            const ghostDistance = this.calculateGhostDanger(
                { x: nextX, y: nextY }, 
                ghostPositions
            );
            const escapeRoutes = this.countEscapeRoutes(nextX, nextY, maze);
            const hasPowerPellet = maze.getCellType(nextX, nextY) === CellType.POWER_PELLET;

            const safetyScore = (ghostDistance * 2.5) + 
                              (escapeRoutes * 1.5) + 
                              (hasPowerPellet ? 8 : 0) +
                              (this.isOppositeDirection(direction) ? -2 : 0);

            if (safetyScore > maxSafety) {
                maxSafety = safetyScore;
                bestDirection = direction;
            }
        }

        return bestDirection;
    }

    isOppositeDirection(newDir) {
        return newDir.x === -this.direction.x && newDir.y === -this.direction.y;
    }

    countEscapeRoutes(x, y, maze) {
        return Object.values(Direction).filter(dir => 
            maze.isValidPosition(x + dir.x, y + dir.y)
        ).length;
    }

    findAlternativePath(maze, ghostPositions) {
        const currentPos = { x: Math.round(this.x), y: Math.round(this.y) };
        const validMoves = Object.values(Direction).filter(dir => {
            const nextX = currentPos.x + dir.x;
            const nextY = currentPos.y + dir.y;
            return maze.isValidPosition(nextX, nextY);
        });

        const safestMove = validMoves.reduce((best, move) => {
            const nextX = currentPos.x + move.x;
            const nextY = currentPos.y + move.y;
            const ghostDanger = this.calculateGhostDanger(
                { x: nextX, y: nextY }, 
                ghostPositions
            );
            
            if (!best || ghostDanger > best.danger) {
                return { move, danger: ghostDanger };
            }
            return best;
        }, null);

        return safestMove ? safestMove.move : this.direction;
    }

    isNearTarget() {
        if (!this.currentPath.length) return true;
        
        const target = this.currentPath[0];
        const distance = Math.sqrt(
            Math.pow(this.x - target.x, 2) + 
            Math.pow(this.y - target.y, 2)
        );
        return distance < 0.3;
    }

    update(maze, ghostPositions) {
        const currentPos = { x: Math.round(this.x), y: Math.round(this.y) };
        let nextDirection = this.direction;

        if (this.autonomousMode) {
            if (this.isStuck()) {
                nextDirection = this.findAlternativePath(maze, ghostPositions);
                this.pathUpdateCooldown = 0;
                this.currentPath = [];
                console.debug('Pacman unstuck: new direction', nextDirection);
            } else {
                const ghostDanger = this.calculateGhostDanger(currentPos, ghostPositions);
                const inDanger = ghostDanger < this.dangerThreshold;

                if (this.pathUpdateCooldown <= 0) {
                    if (inDanger && !this.isPoweredUp) {
                        nextDirection = this.findEscapeDirection(currentPos, ghostPositions, maze);
                        this.currentPath = [];
                        this.pathUpdateCooldown = 2;
                    } else {
                        if (!this.currentPath.length || this.isNearTarget()) {
                            const pellets = this.getPelletPositions(maze);
                            const bestPellet = this.findBestPellet(pellets, ghostPositions, maze);
                            
                            if (bestPellet) {
                                this.currentPath = this.searchAlgorithm.findPath(
                                    currentPos,
                                    [bestPellet],
                                    maze
                                );
                            }
                        }

                        if (this.currentPath.length) {
                            const target = this.currentPath[0];
                            const dx = target.x - this.x;
                            const dy = target.y - this.y;
                            
                            if (Math.abs(dx) > Math.abs(dy)) {
                                nextDirection = dx > 0 ? Direction.RIGHT : Direction.LEFT;
                            } else {
                                nextDirection = dy > 0 ? Direction.DOWN : Direction.UP;
                            }
                        }
                        this.pathUpdateCooldown = 5;
                    }
                } else {
                    this.pathUpdateCooldown--;
                }
            }
        }

        const nextX = this.x + nextDirection.x * this.speed;
        const nextY = this.y + nextDirection.y * this.speed;
        const roundedX = Math.round(nextX);
        const roundedY = Math.round(nextY);

        if (maze.isValidPosition(roundedX, roundedY)) {
            if (!this.autonomousMode || this.isPoweredUp || 
                !this.isUnsafePosition(roundedX, roundedY, ghostPositions)) {
                this.x = nextX;
                this.y = nextY;
                this.direction = nextDirection;

                if (this.currentPath.length && this.isNearTarget()) {
                    this.currentPath.shift();
                }

                if (Math.abs(this.x - roundedX) < 0.3 && Math.abs(this.y - roundedY) < 0.3) {
                    const isPowerPellet = maze.eatPellet(roundedX, roundedY);
                    if (isPowerPellet) {
                        this.isPoweredUp = true;
                        this.powerTimer = 600;
                        this.score += 50;
                    } else {
                        this.score += 10;
                    }
                }
            }
        }

        if (this.isPoweredUp) {
            this.powerTimer--;
            if (this.powerTimer <= 0) {
                this.isPoweredUp = false;
            }
        }

        this.updateMouthAnimation();
    }

    isUnsafePosition(x, y, ghostPositions) {
        return ghostPositions.some(ghost => 
            Math.abs(x - ghost.x) + Math.abs(y - ghost.y) < 2
        );
    }

    draw(ctx, offset = 0) {
        const centerX = Math.round(this.x * CELL_SIZE + CELL_SIZE / 2);
        const centerY = Math.round(this.y * CELL_SIZE + CELL_SIZE / 2) + offset;
        const radius = Math.round(CELL_SIZE * 0.8 / 2);

        let rotationAngle = 0;
        if (this.direction === Direction.UP) rotationAngle = -90;
        else if (this.direction === Direction.DOWN) rotationAngle = 90;
        else if (this.direction === Direction.LEFT) rotationAngle = 180;

        ctx.save();
        ctx.translate(centerX, centerY);
        ctx.rotate(rotationAngle * Math.PI / 180);
        ctx.translate(-centerX, -centerY);

        ctx.beginPath();
        ctx.fillStyle = this.isPoweredUp ? 
            `hsl(60, 100%, ${50 + Math.sin(Date.now() / 100) * 20}%)` : 
            this.color;
        
        ctx.arc(
            centerX, 
            centerY, 
            radius,
            (this.mouthAngle * Math.PI / 180),
            (360 - this.mouthAngle) * Math.PI / 180
        );
        ctx.lineTo(centerX, centerY);
        ctx.fill();

        if (this.autonomousMode && this.currentPath.length) {
            ctx.restore();
            ctx.strokeStyle = 'rgba(255, 0, 0, 0.5)';
            ctx.lineWidth = 2;
            ctx.beginPath();
            ctx.moveTo(centerX, centerY);
            
            this.currentPath.forEach(pos => {
                const pathX = pos.x * CELL_SIZE + CELL_SIZE / 2;
                const pathY = pos.y * CELL_SIZE + CELL_SIZE / 2 + offset;
                ctx.lineTo(pathX, pathY);
            });
            
            ctx.stroke();
            return;
        }

        ctx.restore();
    }

    updateMouthAnimation() {
        if (this.openingMouth) {
            this.mouthAngle = Math.min(45, this.mouthAngle + this.mouthSpeed);
            if (this.mouthAngle >= 45) this.openingMouth = false;
        } else {
            this.mouthAngle = Math.max(0, this.mouthAngle - this.mouthSpeed);
            if (this.mouthAngle <= 0) this.openingMouth = true;
        }
    }

    toggleControlMode() {
        this.autonomousMode = !this.autonomousMode;
        this.currentPath = [];
        this.pathUpdateCooldown = 0;
        this.noMovementCounter = 0;
        console.log(`Control mode: ${this.autonomousMode ? 'AI' : 'Manual'}`);
    }

    getPosition() {
        return { x: Math.round(this.x), y: Math.round(this.y) };
    }
}